### ISSUE TITLE
zfs module somtimes changes properties where no changes necessary

### ISSUE THREAD

**Issue by froebela** [ansible/ansible#66502]
##### SUMMARY

In ZFS locally set properties of a certain ZFS filesystem might actually come from two different sources, local or received. Local means, this property was explicitly set using zfs set. Received means, this property was implicitly set by a zfs send/receive operation, for example during a template based Solaris installation. Hence the Ansible ZFS module should also consider properties of source "received" as "local".

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

lib/ansible/modules/storage/zfs/zfs.py
##### ANSIBLE VERSION

```paste below
ansible 2.9.9
  config file = /adm/afroebel/ansible/unix/ansible.cfg
  configured module search path = [u'/adm/afroebel/ansible/unix/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.16 (default, Oct 10 2019, 22:02:15) [GCC 8.3.0]
```

##### CONFIGURATION

```paste below
COLOR_VERBOSE(/adm/afroebel/ansible/unix/ansible.cfg) = cyan
DEFAULT_GATHERING(/adm/afroebel/ansible/unix/ansible.cfg) = explicit
DEFAULT_MANAGED_STR(/adm/afroebel/ansible/unix/ansible.cfg) = {file}
DEFAULT_PRIVATE_ROLE_VARS(/adm/afroebel/ansible/unix/ansible.cfg) = True
PERSISTENT_CONNECT_TIMEOUT(/adm/afroebel/ansible/unix/ansible.cfg) = 30
```

##### OS / ENVIRONMENT

Ansible controller: Debian GNU/Linux 10
managed devices: Oracle Solaris 11.3/11.4

##### STEPS TO REPRODUCE

We've got a Solaris system with 2 different ZFS filesystems:

```yaml
root@sap15dbz1# zfs get -r mountpoint rpool/export/home/oracle rpool/export/home/z15adm
NAME                      PROPERTY    VALUE                SOURCE
rpool/export/home/oracle  mountpoint  /export/home/oracle  received
rpool/export/home/z15adm  mountpoint  /export/home/z15adm  local
```
As one can see, both ZFS filesystems have a locally set property "mountpoint" comming from different sources. ZFS filesystem rpool/export/home/oracle was created during system installation and ZFS filesystem rpool/export/home/z15adm was manually created afterwards.

We wrote a short playbook using the zfs module in order to set the ZFS property mountpoint to the current values.
```yaml
# cat test/zfs.yml
---
- name:   "Test Playbook"
  hosts:  "sap15dbz1"
  tasks:

    - name: "Set ZFS property mountpoint for fs rpool/export/home/oracle"
      zfs:
        name:  "rpool/export/home/oracle"
        state: present
        extra_zfs_properties:
          mountpoint: "/export/home/oracle"

    - name: "Set ZFS property mountpoint for fs rpool/export/home/z15adm"
      zfs:
        name:  "rpool/export/home/z15adm"
        state: present
        extra_zfs_properties:
          mountpoint: "/export/home/z15adm"
```
 

##### EXPECTED RESULTS

As the values for ZFS property "mountpoint" in the Ansible playbook were already equal to the current property values on the managed system, we would expect the zfs module to not detect any changes.

##### ACTUAL RESULTS

Instead it detected changes for the property of source "received".

```yaml
# ansible-playbook -i inventory test/zfs.yml -C -D

PLAY [Test Playbook] *******************************************************************

TASK [Set ZFS property mountpoint for fs rpool/export/home/oracle] **********************
changed: [sap15dbz1]

TASK [Set ZFS property mountpoint for fs rpool/export/home/z15adm] *********************
ok: [sap15dbz1]

PLAY RECAP *****************************************************************************
sap15dbz1:  ok=2  changed=1  unreachable=0  failed=0  skipped=0  rescued=0  ignored=0
```
Now one could argue, that setting a ZFS property once more to the current value doesn't hurt much. But first detecting unnecessary changes is confusing (diff mode support to qualify them would be nice) and second in older Solaris releases setting the ZFS property mountpoint always led to an umount/remount operation of the affected ZFS filesystem and in most cases couldn't be done online. Therefore we would really apprechiate, if this could be fixed/enhanced soon.

As far as we've seen in the code, this could be easily accomplished by changing a single line in method get_current_properties from _if source == 'local':_ to _if source == 'local' or source == 'received':_.

**Comment by ansibot** [comment]
Files identified in the description:
* [`lib/ansible/modules/storage/zfs/zfs.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/storage/zfs/zfs.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @bcoca @dagwieers @danowar2k @fishman @jasperla @johanwiren @jpdasma @mator @scathatheworm @troy2914 @xen0l
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by froebela** [comment]
I created pull request #69964 in order to address it for Ansible release 2.9.

**Comment by mkrizek** [comment]
The change would need to be merged in https://github.com/ansible-collections/community.general/ first in order to be backported to 2.9 in ansible/ansible.

**Comment by froebela** [comment]
Thanks for pointing that out. To be honest, I wanted to open a pull request in the devel branch of the [ansible/ansible](https://github.com/ansible/ansible) repo, but couldn't find the zfs module anymore. As I didn't know at that time, where it has gone, I decided to open the pull request for the stable-2.9 branch instead.

As I'm quite new to contributing to the Ansible code I need some help in how to proceed. How do I get my change to be merged into [ansible-collections/community.general](https://github.com/ansible-collections/community.general/)? Do I need to take care of it by myself? If yes, do I need to cancel my original pull request for ansible/ansible:stable-2.9 and create a new one for that repo or is there a way to somehow transfer it?

**Comment by mkrizek** [comment]
You need to create a new pull request manually in https://github.com/ansible-collections/community.general. Once that is merged, open a backport PR in ansible/ansible.

**Comment by froebela** [comment]
I updated the problem description in order to reflect our current Ansible versions. The issue also exists in Ansible 2.9 in exactly the same way as it did for Ansible 2.8.

Furthermore I created pull request [#502](https://github.com/ansible-collections/community.general/pull/502) for [ansible-collections/community.general:master](https://github.com/ansible-collections/community.general), but left pull request [#69964](https://github.com/ansible/ansible/pull/69964) for [https://github.com/ansible/ansible:stable-2.9](https://github.com/ansible/ansible/tree/stable-2.9) open for backporting. Hope this was ok, otherwise please tell me, what to change.

**Comment by ansibot** [comment]
Thank you very much for your interest in Ansible. Ansible has migrated much of the content into separate repositories to allow for more rapid, independent development. We are closing this issue/PR because this content has been moved to one or more collection repositories.

* lib/ansible/modules/storage/zfs/zfs.py -> https://galaxy.ansible.com/community/general

For further information, please see:
https://github.com/ansible/ansibullbot/blob/master/docs/collection_migration.md
